### ISSUE TITLE
92% chunk asset optimization TerserPlugin -JavaScript heap out of memory

### ISSUE THREAD

**Issue by elouizbadr** [angular/angular-cli#13939]
# 🐞 Bug report

### Command (mark with an `x`)

```
- [ ] new
- [x] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Is this a regression?

 No

### Description

 A clear and concise description of the problem...

## 🔥 Exception or Error
<pre><code>

> ng-x-rocket@1.0.0 build /home/ec2-user/environment/ziwig/client/web
> npm run env -s && ng build --prod --configuration=$STAGE

 92% chunk asset optimization TerserPlugin                                         
<--- Last few GCs --->

[1577:0x3650560]   334791 ms: Mark-sweep 1368.0 (1531.2) -> 1367.9 (1531.2) MB, 1474.2 / 0.0 ms  allocation failure GC in old space requested
[1577:0x3650560]   336881 ms: Mark-sweep 1367.9 (1531.2) -> 1367.7 (1486.2) MB, 2090.2 / 0.0 ms  last resort GC in old space requested
[1577:0x3650560]   338694 ms: Mark-sweep 1367.7 (1486.2) -> 1367.7 (1480.2) MB, 1812.0 / 0.0 ms  last resort GC in old space requested

<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x104e83d25879 <JSObject>
    2: /* anonymous */(aka /* anonymous */) [/home/ec2-user/environment/ziwig/client/web/node_modules/terser/dist/bundle.min.js:1] [bytecode=0x301009147a29 offset=87](this=0x3e74bea022d1 <undefined>,n=0x336ec76c57e9 <yn map = 0xac305745019>)
    3: arguments adaptor frame: 2->1
    4: each [/home/ec2-user/environment/ziwig/client/web/node_modules/terser/dist/bundle.min.js:~1] [pc=0x22698f949d82]...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [ng]
 2: 0x8cbf4c [ng]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [ng]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [ng]
 5: v8::internal::Factory::NewUninitializedFixedArray(int) [ng]
 6: 0xd801bc [ng]
 7: 0xd97a95 [ng]
 8: v8::internal::JSObject::AddDataElement(v8::internal::Handle<v8::internal::JSObject>, unsigned int, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::Object::ShouldThrow) [ng]
 9: v8::internal::Object::AddDataProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::Object::ShouldThrow, v8::internal::Object::StoreFromKeyed) [ng]
10: v8::internal::Object::SetProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::LanguageMode, v8::internal::Object::StoreFromKeyed) [ng]
11: v8::internal::Runtime_SetProperty(int, v8::internal::Object**, v8::internal::Isolate*) [ng]
12: 0x22698de042fd
sh: line 1:  1577 Aborted                 ng build --prod --configuration=$STAGE
npm ERR! code ELIFECYCLE
npm ERR! errno 134
npm ERR! ng-x-rocket@1.0.0 build: `npm run env -s && ng build --prod --configuration=$STAGE`
npm ERR! Exit status 134
npm ERR! 
npm ERR! Failed at the ng-x-rocket@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ec2-user/.npm/_logs/2019-03-18T09_38_54_547Z-debug.log
</code></pre>
## 🌍 Your Environment
<pre><code>

Angular CLI: 6.1.5
Node: 8.12.0
OS: linux x64
Angular: 6.1.10
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router, service-worker

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.13.6
@angular-devkit/build-angular     0.13.6
@angular-devkit/build-optimizer   0.13.6
@angular-devkit/build-webpack     0.13.6
@angular-devkit/core              7.3.6
@angular-devkit/schematics        0.7.5
@angular/cli                      6.1.5
@ngtools/webpack                  7.3.6
@schematics/angular               0.7.5
@schematics/update                0.7.5
rxjs                              6.3.3
typescript                        2.7.2
webpack                           4.29.0
</code></pre>
**Anything else relevant?**
- node version : 8.12.0
- npm version : 6.4.1
- content of package.json 
<code><pre>
{
    "name": "ng-x-rocket",
    "version": "1.0.0",
    "private": true,
    "scripts": {
        "ng": "ng",
        "build": "npm run env -s && ng build --prod --configuration=$STAGE",
        "start": "npm run env -s && ng serve --aot --proxy-config proxy.conf.js",
        "serve:sw": "npm run build -s && npx http-server ./dist -p 4200",
        "lint": "ng lint && stylelint \"src/**/*.scss\" --syntax scss && htmlhint \"src\" --config .htmlhintrc",
        "test": "npm run env -s && ng test",
        "test:ci": "npm run env -s && npm run lint -s && ng test --code-coverage --watch=false",
        "e2e": "npm run env -s && ng e2e",
        "translations:extract": "ngx-translate-extract --input ./src --output ./src/translations/template.json --format=json --clean -sort --marker extract",
        "docs": "hads ./docs -o",
        "env": "ngx-scripts env npm_package_version",
        "generate": "ng generate"
    },
    "dependencies": {
        "@angular/animations": "~6.1.10",
        "@angular/common": "~6.1.10",
        "@angular/compiler": "~6.1.10",
        "@angular/core": "~6.1.10",
        "@angular/forms": "~6.1.10",
        "@angular/http": "~6.1.10",
        "@angular/platform-browser": "~6.1.10",
        "@angular/platform-browser-dynamic": "~6.1.10",
        "@angular/router": "~6.1.10",
        "@angular/service-worker": "~6.1.10",
        "@fortawesome/fontawesome-free-webfonts": "~1.0.9",
        "@ng-bootstrap/ng-bootstrap": "~2.2.2",
        "@ng-select/ng-select": "~2.15.1",
        "@ngx-rocket/scripts": "~3.0.1",
        "@ngx-translate/core": "~10.0.2",
        "angular-tree-component": "~8.2.0",
        "async": "~2.6.1",
        "aws-sdk": "~2.395.0",
        "bootstrap": "~4.2.1",
        "core-js": "~2.6.3",
        "font-awesome": "~4.7.0",
        "lodash": "~4.17.11",
        "moment": "~2.22.2",
        "ng-circle-progress": "~1.3.0",
        "ng5-slider": "~1.1.12",
        "ngx-bootstrap": "~3.2.0",
        "ngx-color-picker": "~6.8.0",
        "ngx-drag-and-drop-lists": "1.1.3",
        "ngx-editor": "~3.3.0",
        "ngx-moment": "~3.2.0",
        "ngx-owl-carousel-o": "~1.0.5",
        "ngx-toggle-switch": "~2.0.5",
        "rxjs": "~6.3.3",
        "rxjs-compat": "~6.3.3",
        "zone.js": "~0.8.22"
    },
    "devDependencies": {
        "@angular-builders/custom-webpack": "latest",
        "@angular-builders/dev-server": "latest",
        "@angular-devkit/build-angular": "~0.13.5",
        "@angular/cli": "~6.1.5",
        "@angular/compiler-cli": "~6.1.10",
        "@angular/language-service": "~6.1.10",
        "@biesbjerg/ngx-translate-extract": "~2.3.4",
        "@types/jasmine": "~2.8.6",
        "@types/jasminewd2": "~2.0.3",
        "@types/node": "~8.9.4",
        "codelyzer": "~4.2.1",
        "dotenv": "latest",
        "jasmine-core": "~2.99.1",
        "jasmine-spec-reporter": "~4.2.1",
        "karma": "~1.7.1",
        "karma-chrome-launcher": "~2.2.0",
        "karma-coverage-istanbul-reporter": "~2.0.0",
        "karma-jasmine": "~1.1.1",
        "karma-jasmine-html-reporter": "~0.2.2",
        "protractor": "~5.4.0",
        "ts-node": "~5.0.1",
        "tslint": "~5.9.1",
        "typescript": "~2.7.2"
    }
}
</pre></code>
- content of "/home/ec2-user/.npm/_logs/2019-03-18T09_38_54_547Z-debug.log" :
<code><pre>
0 info it worked if it ends with ok
1 verbose cli [ '/home/ec2-user/.nvm/versions/node/v8.12.0/bin/node',
1 verbose cli   '/home/ec2-user/.nvm/versions/node/v8.12.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'build' ]
2 info using npm@6.4.1
3 info using node@v8.12.0
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle ng-x-rocket@1.0.0~prebuild: ng-x-rocket@1.0.0
6 info lifecycle ng-x-rocket@1.0.0~build: ng-x-rocket@1.0.0
7 verbose lifecycle ng-x-rocket@1.0.0~build: unsafe-perm in lifecycle true
8 verbose lifecycle ng-x-rocket@1.0.0~build: PATH: /home/ec2-user/.nvm/versions/node/v8.12.0/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/ec2-user/environment/ziwig/client/web/node_modules/.bin:/home/ec2-user/.nvm/versions/node/v8.12.0/bin:/usr/local/rvm/gems/ruby-2.4.1/bin:/usr/local/rvm/gems/ruby-2.4.1@global/bin:/usr/local/rvm/rubies/ruby-2.4.1/bin:/usr/local/bin:/bin:/usr/bin:/home/ec2-user/.local/bin:/home/ec2-user/bin:/usr/local/sbin:/usr/sbin:/sbin:/opt/aws/bin:/usr/local/rvm/bin:/home/ec2-user/.local/bin:/home/ec2-user/bin:/home/ec2-user/.local/bin:/home/ec2-user/bin
9 verbose lifecycle ng-x-rocket@1.0.0~build: CWD: /home/ec2-user/environment/ziwig/client/web
10 silly lifecycle ng-x-rocket@1.0.0~build: Args: [ '-c',
10 silly lifecycle   'npm run env -s && ng build --prod --configuration=$STAGE' ]
11 silly lifecycle ng-x-rocket@1.0.0~build: Returned: code: 134  signal: null
12 info lifecycle ng-x-rocket@1.0.0~build: Failed to exec build script
13 verbose stack Error: ng-x-rocket@1.0.0 build: `npm run env -s && ng build --prod --configuration=$STAGE`
13 verbose stack Exit status 134
13 verbose stack     at EventEmitter.<anonymous> (/home/ec2-user/.nvm/versions/node/v8.12.0/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)
13 verbose stack     at emitTwo (events.js:126:13)
13 verbose stack     at EventEmitter.emit (events.js:214:7)
13 verbose stack     at ChildProcess.<anonymous> (/home/ec2-user/.nvm/versions/node/v8.12.0/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at emitTwo (events.js:126:13)
13 verbose stack     at ChildProcess.emit (events.js:214:7)
13 verbose stack     at maybeClose (internal/child_process.js:915:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)
14 verbose pkgid ng-x-rocket@1.0.0
15 verbose cwd /home/ec2-user/environment/ziwig/client/web
16 verbose Linux 4.14.47-56.37.amzn1.x86_64
17 verbose argv "/home/ec2-user/.nvm/versions/node/v8.12.0/bin/node" "/home/ec2-user/.nvm/versions/node/v8.12.0/bin/npm" "run" "build"
18 verbose node v8.12.0
19 verbose npm  v6.4.1
20 error code ELIFECYCLE
21 error errno 134
22 error ng-x-rocket@1.0.0 build: `npm run env -s && ng build --prod --configuration=$STAGE`
22 error Exit status 134
23 error Failed at the ng-x-rocket@1.0.0 build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 134, true ]
</pre></code>

**Comment by alan-agius4** [comment]
Hi, does increasing the memory limit help? See: https://github.com/angular/angular-cli/issues/13853#issuecomment-470833135

How big is your project?

**Comment by elouizbadr** [comment]
@alan-agius4 I guess it's a medium project. Thanks for the reference because this problem was fixed by https://github.com/angular/angular-cli/issues/5618#issuecomment-450151214. Meanwhile, this should not be happening because it shouldn't take all that `node --max_old_space_size=2048` amount of memory to build the project.

**Comment by rip222** [comment]
> @alan-agius4 I guess it's a medium project.
> Thanks for the reference because this problem was fixed by [#5618 (comment)](https://github.com/angular/angular-cli/issues/5618#issuecomment-450151214).

that didn't help me :(

**Comment by progmars** [comment]
The same issue. 

I've already set the limit and verified that it works when you have enough memory, but still the amount required is ridiculously high.

Windows:
```
set NODE_OPTIONS="--max-old-space-size=8192" && node -e "(() => { console.log(require('v8').getHeapStatistics()); })();"
```

on an Ubuntu VM and inside Docker container:
```
export NODE_OPTIONS=--max-old-space-size=8192 && node -e "(() => { console.log(require('v8').getHeapStatistics()); })();"
```
Important: no quotes around the variable and - not _

It reports:
```
  total_available_size: 8762699648,
  heap_size_limit: 8769069275,
```

so, I use the same commands to run my `ng build --prod`:
on Windows:
```
set NODE_OPTIONS=--max-old-space-size=8192 && ng build --prod
```

on an Ubuntu VM and inside Docker container:
```
set NODE_OPTIONS=--max-old-space-size=8192 && ng build --prod
```

It works fine on the Windows machine because it has 16 GB of RAM.

But it fails on Ubuntu with 5 GB of physical RAM and 1 GB swap (oops, too little swap, will have to ask our admin to increase it).

Anyway, 5 GB of RAM for compiling a not-very-complex web app seems to be somewhat too high.
